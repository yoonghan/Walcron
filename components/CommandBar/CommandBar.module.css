.container {
  position: relative;
}

.header {
  text-align: center;
  position: relative;
  background-color: #fff;
}

.header.commandPromptOnly {
  left: 0;
  right: 0;
}

.desktop.shift-to-right {
  position: relative;
  display: flex;
  flex-direction: row-reverse;
  justify-content: space-between;
  padding: 0 1rem;
}

.desktop {
  display: block;
}

.mobile {
  display: none;
}

.loading {
  font-family: "Inconsolata";
  color: green;
}

@media only screen and (max-width: 480px) {
  .mobile {
    display: block;
  }
  .desktop {
    display: none !important;
  }
}
