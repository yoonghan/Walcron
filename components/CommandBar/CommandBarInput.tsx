`use strict`

/**
  A command prompt input
  **/

import * as React from "react";

export interface CommandBarInputProps {
  onBlurCallback: (event :React.FormEvent<HTMLInputElement>)=>void;
  onSubmitCallback: (event :React.FormEvent<HTMLFormElement>)=>void;
  isPromptWait: boolean;
}

const CommandBarInput: React.SFC<CommandBarInputProps> = ({onBlurCallback, onSubmitCallback, isPromptWait}) => {
  function _onClickSelect(event: React.FormEvent<HTMLInputElement>) {
    (event.target as any).select();
  }

  return (
    <form onSubmit={onSubmitCallback} className="command-container">
      <div className="command-text-container">
        <div className="command-text-prompt">
          guest@walcron$
        </div>
        <input
          id="command-text"
          autoCapitalize="none"
          autoComplete="off"
          autoCorrect="off"
          tabIndex={0}
          type="text"
          spellCheck={false}
          aria-label="command prompt"
          aria-haspopup="false"
          aria-autocomplete="list"
          dir="ltr"
          onBlur={onBlurCallback}
          onClick={_onClickSelect}
          />
        {isPromptWait && <span>&#9608;</span>}
      </div>
      <button id="command-enter" className="style-scope ytd-searchbox" aria-label="Enter">
        <i className="fas fa-arrow-right"></i>
      </button>
      <style jsx>{`
        .command-container {
          display: flex;
          font-size: 0.8rem;
          position: relative;
        }
        .command-text-container	span {
          color: #FFF;
        	position: absolute;
        	width: 1px;
          height: 100%;
        	left: 7rem;
          top: 0.4rem;
        	animation-name: blink;
        	animation-duration: 800ms;
        	animation-iteration-count: infinite;
        	opacity: 1;
        }
        .command-text-container input:focus + span {
        	display: none;
        }
        @keyframes blink {
        	from { opacity: 1; }
        	to { opacity: 0; }
        }
        .command-text-prompt {
          width: 7rem;
          height: 100%;
          display: flex;
          position: absolute;
          align-items: center;
          pointer-events: none;
          justify-content: center;
          color: #FFF;
          font-family: Inconsolata;
        }
        #command-text {
          font-family: Inconsolata;
          color: #EBEBEB;
          background-color: #343436;
          width: 250px;
          border: 1px solid;
          margin: 0;
          display: block;
          min-width: 0;
          font-size: 0.8rem;
          box-sizing: content-box;
          -webkit-tap-highlight-color: transparent;
          padding: 0.5rem 2rem 0.5rem 7rem;
          border-radius: 0.3rem;
        }
        #command-text::placeholder {
          color: #A2A2A2;
        }
        #command-enter {
          position: absolute;
          height: 100%;
          right: 3px;
          transition-property: color, background;
          transition-duration: .15s;
          transition-timing-function: ease-in-out;
          color: #FFF;
          border: none;
          background: none;
          font-size: 0.8rem;
          user-select: none;
          cursor: pointer;
        }
        #command-enter:hover {
          color: #000;
        }
        #command-enter:active {
          margin: 1px 0 0 1px;
        }
      `}</style>
    </form>
  );
}

export default CommandBarInput;
