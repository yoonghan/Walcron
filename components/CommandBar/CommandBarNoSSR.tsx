import * as React from "react"
import NavMenu from "./NavMenu"
import Logo from "../Logo"
import NoSSRCommandBar from "./NoSSRCommandBar"
import NoSSRMobileMenu from "./NoSSRMobileMenu"
import styles from "./CommandBar.module.css"

export interface CommandBarNoSSRProps {
  disableMobile?: boolean
  commandPromptOnly?: boolean
}

const CommandBarNoSSR = ({
  disableMobile,
  commandPromptOnly,
}: CommandBarNoSSRProps) => {
  return (
    <div
      className={`${styles["header"]} ${
        commandPromptOnly ? styles["commandPromptOnly"] : ""
      }`}
      id="commandbar"
    >
      <div
        className={
          (!commandPromptOnly ? styles["desktop"] : "") +
          " " +
          (!disableMobile && !commandPromptOnly ? styles["shift-to-right"] : "")
        }
      >
        <NoSSRCommandBar />
        {!disableMobile && !commandPromptOnly && <NavMenu />}
      </div>
      {!disableMobile && !commandPromptOnly && (
        <div className={styles["mobile"]}>
          <NoSSRMobileMenu />
        </div>
      )}
      {disableMobile && <Logo />}
    </div>
  )
}

export default CommandBarNoSSR
