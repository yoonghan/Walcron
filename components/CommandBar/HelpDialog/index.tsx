`use strict`

/**
  Displaying the dialog for help
  **/

import * as React from "react";
import Modal from "../../Modal";
import HelpContent from "./content";

interface HelpDialogProps {
  cancelCallback: () => void;
}

const HelpDialog: React.SFC<HelpDialogProps> = ({cancelCallback}) => (
  <Modal cancelCallback={cancelCallback}>
    <HelpContent/>
  </Modal>
)

export default HelpDialog;
