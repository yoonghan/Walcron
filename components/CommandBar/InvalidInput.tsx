import * as React from "react";
import {ERROR} from "../../shared/style";

interface InvalidInputProps {
  invalidInput: string;
}

const trimInput = (input:string) => {
  const _input = input.trim();
  return _input.length > 5 ? _input.substring(0,5)+"...": _input;
}

const InvalidInput: React.SFC<InvalidInputProps> = ({invalidInput}) => {
  return (
    <div className="error">
      {trimInput(invalidInput)} - command not found
      <style jsx>{`
        .error {
          font-family: Inconsolata;
          font-size: ${ERROR.FONT_SIZE};
          color: ${ERROR.FOREGROUND};
          position: absolute;
          top: 0;
          left: 7rem;
        }
      `}</style>
    </div>
  );
}

export default InvalidInput;
