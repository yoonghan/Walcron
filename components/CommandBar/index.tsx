import * as React from "react";
import NoSSR from 'react-no-ssr';
import NavMenu from "./NavMenu";
import NoSSRCommandBar from "./NoSSRCommandBar";
import NoSSRMobileMenu from "./NoSSRMobileMenu";

interface CommandBarProps {
  disableMobile?: boolean;
}

const CommandBar: React.SFC<CommandBarProps> = ({disableMobile}) => {
  return (
    <NoSSR>
      <div className={"header"} id="commandbar">
        <div className={"desktop" + (!disableMobile? " shift-to-right":"")}>
          <NoSSRCommandBar/>
          {!disableMobile && <NavMenu/>}
        </div>
        {!disableMobile && (<div className="mobile">
          <NoSSRMobileMenu/>
        </div>)}
        <style jsx>{`
          .header {
            text-align: center;
            position: absolute;
            top: 0.5rem;
            left: 1.5rem;
            z-index: 2;
            right: 1.5rem;
          }
          .desktop.shift-to-right {
            position: relative;
            display: flex;
            flex-direction: row-reverse;
            justify-content: space-between;
          }
          .desktop {
            display: block;
          }
          .mobile {
            display: none;
          }
          @media only screen and (max-width: 480px) {
            .mobile {
              display: block;
            }
            .desktop {
              display: none !important;
            }
          }
        `}</style>
      </div>
    </NoSSR>
  );
}

export default CommandBar;
