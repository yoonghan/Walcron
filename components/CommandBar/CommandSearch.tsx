import * as React from "react";
import ReactDOM from 'react-dom';
import { RouterProps } from "next-server/router";
import HelpDialog from "./HelpDialog";
import InvalidCommand from "./InvalidCommand";
import Output from "./Output";

export enum EnumAction {
  COMMAND,
  LINK
}

const ABOUT_LINK = {
  action: EnumAction.LINK,
  url: "/about",
  exec: (router: RouterProps) => {
    if(router.route === "/about") {
      return <InvalidCommand invalidCommand={"This is the page"}/>;
    }
    router.push("/about");
    return <React.Fragment/>;
  }
}

export interface ICommand {
  [s: string]: IAvailableInput;
}

export interface IAvailableInput {
  action: EnumAction;
  description: string;
  url?: string;
  exec: any;
  synonym?: Array<string>;
}

export const AvailableInput:ICommand = {
  "whoami": {
    synonym: ["cd about", "cd /about"],
    description: "Get to know this site.",
    ...ABOUT_LINK
  },
  "su - walcron": {
    synonym: ["sudo su", "sudo walcron", "su -", "sudo su -"],
    description: "Know us by being us.",
    ...ABOUT_LINK
  },
  "pwd": {
    description: "Lost, and need direction.",
    action: EnumAction.LINK,
    exec: (router: RouterProps) => {
      return <Output output={router.route}/>;
    }
  },
  "ls": {
    synonym: ["dir", "cd creation", "cd /creation", "cd /invent"],
    description: "What's there ?",
    action: EnumAction.LINK,
    exec: (router: RouterProps) => {
      if(router.route === "/creation") {
        return <InvalidCommand invalidCommand={"This is the page"}/>;
      }
      router.push("/creation");
      return <React.Fragment/>;
    }
  },
  "exit": {
    description: "Return to main page",
    action: EnumAction.LINK,
    exec: (router: RouterProps) => {
      if(router.route === "/") {
        return <InvalidCommand invalidCommand={"Already at root"}/>;
      }
      location.href="/";
      return <React.Fragment/>;
    }
  },
  "cd ..": {
    description: "Return to previous page",
    action: EnumAction.LINK,
    exec: (router: RouterProps) => {
      if(router.route !== "/") {
        router.back();
        return <React.Fragment/>
      }
      else
        return <InvalidCommand invalidCommand={"Already at root"}/>
    }
  },
  "cd /": {
    synonym: ["cd"],
    description: "Return to main page",
    action: EnumAction.LINK,
    exec: (router: RouterProps) => {
      if(router.route === "/") {
        return <InvalidCommand invalidCommand={"Already at root"}/>;
      }
      router.push("/");
      return <React.Fragment/>;
    }
  },
  "=": {
    description: "Do Math",
    action: EnumAction.COMMAND,
    exec: () => <InvalidCommand invalidCommand={"Provide an equation"}/>
  },
  "help": {
    description: "Lost, confused, need help.",
    synonym: ["man", "info"],
    action: EnumAction.COMMAND,
    exec: (element:HTMLDivElement, cancellationCallback:()=>void) => {
      return ReactDOM.createPortal(
        <HelpDialog cancelCallback={cancellationCallback}/>,
        element
      );
    }
  }
};
