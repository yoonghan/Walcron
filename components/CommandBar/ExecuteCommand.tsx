import * as React from "react";
import InvalidInput from "./InvalidInput";
import InvalidCommand from "./InvalidCommand";
import Output from "./Output";
import {AvailableInput, IAvailableInput, EnumAction} from "./CommandSearch";
import * as ts from "typescript";

interface ExtendedRouterProps {
}


function evaluateMath(mathEval: string):string {
  // const code: string = `({
  //   Run: (data: string): string => {
  //       console.log(data); return Promise.resolve("SUCCESS"); }
  //   })`;

  const result = ts.transpile(mathEval);
  const evaluatedResult :any = eval(result);
  // runnalbe.Run("RUN!").then((result:string)=>{console.log(result);});
  return evaluatedResult;
}

function getMathEvaluation(evaluation: string) {
  const equalsLocation = 1;
  const _evaluation = evaluation.replace(/ /g,'').substr(equalsLocation);

  const mathRegex = /^[0-9]+(\.[0-9]*)?((\+|\-|\*|\/)[0-9]+(\.[0-9]*)?)+$/;
  const matches = mathRegex.test(_evaluation);

  if(matches) {
    const result = evaluateMath(_evaluation);
    return <Output output={result}/>
  }
  else {
    return <InvalidCommand invalidCommand={"Unable to evaluate."}/>;
  }
}

function isSpecialCommand(inputCommand:string) {
  return (inputCommand.indexOf('=') === 0 && inputCommand.length > 3);
}

export function exec(element:HTMLDivElement, cancellationCallback:()=>void, router:ExtendedRouterProps) {

  const executeBasedOnType = (inputCommand:string, inputSelected: IAvailableInput) => {
    switch(inputSelected.action) {
      case EnumAction.COMMAND:
        return executeCommand(inputCommand);
      case EnumAction.LINK:
        return executeLink(inputCommand);
      default:
        return <React.Fragment/>;
    }
  }

  const executeCommand = (inputCommand:string) => {
    return AvailableInput[inputCommand].exec(element, cancellationCallback);
  }

  const executeLink = (inputCommand:string) => {
    return AvailableInput[inputCommand].exec(router);
  }

  return function(inputCommand:string) {
    const _inputCommand = inputCommand.trim().toLowerCase();
    const matchedCommand:IAvailableInput = AvailableInput[_inputCommand];
    if(matchedCommand) {
      return executeBasedOnType(_inputCommand, matchedCommand);
    }
    else if(isSpecialCommand(inputCommand)) {
      return getMathEvaluation(inputCommand);
    }
    else {
      return <InvalidInput invalidInput={inputCommand}/>
    }
  }
}
