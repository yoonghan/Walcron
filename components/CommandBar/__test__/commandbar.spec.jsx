import * as React from 'react';
import { shallow, mount } from 'enzyme';
import CommandBar from "../index";

//Increase wait time
jest.setTimeout(5000);

//Rewrite router.
jest.mock("next/router");

//Rewrite dynamic.
jest.mock('next/dynamic');

const waitNextTick = async () => {
  return new Promise(resolve => setTimeout(resolve, 2000));
}

const mountInitialComponent = async (component) => {
  const wrapper = mount(component);
  await waitNextTick();
  wrapper.update();
  return wrapper;
};

describe('CommandBar', () => {
  it('renders lazy normally', () => {
    const wrapper = shallow(<CommandBar/>);
    expect(wrapper.debug()).toMatchSnapshot();
  }),
  it('renders lazy normally', async () => {
    const wrapper = await mountInitialComponent(<CommandBar/>);
    expect(wrapper.debug()).toMatchSnapshot();
  })
});
