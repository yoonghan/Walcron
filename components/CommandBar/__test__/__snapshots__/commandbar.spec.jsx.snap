// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandBar renders lazy normally 1`] = `
"<NoSSR>
  <Suspense fallback={{...}}>
    <lazy />
  </Suspense>
</NoSSR>"
`;

exports[`CommandBar renders lazy normally 2`] = `
"<Memo(CommandBar)>
  <NoSSR>
    <Suspense fallback={{...}}>
      <CommandBarNoSSR>
        <div id=\\"commandbar\\" className=\\"jsx-3778838050 header\\">
          <div className=\\"jsx-3778838050 desktop shift-to-right\\">
            <NoSSRCommandBar router={{...}}>
              <div className=\\"jsx-2638007119 container\\">
                <CommandBarInput onBlurCallback={[Function (anonymous)]} onFocusCallback={[Function (anonymous)]} onSubmitCallback={[Function (anonymous)]}>
                  <form onSubmit={[Function (anonymous)]} className=\\"jsx-2334938183 command-container\\">
                    <div className=\\"jsx-2334938183 command-text-container\\">
                      <div className=\\"jsx-2334938183 command-text-prompt\\">
                        walcron@tm$
                      </div>
                      <div className=\\"jsx-2334938183 prompt\\">
                        <span className=\\"jsx-2334938183 promptIn\\" />
                        <span />
                      </div>
                      <Autosuggest suggestions={{...}} onSuggestionsFetchRequested={[Function (anonymous)]} onSuggestionsClearRequested={[Function (anonymous)]} getSuggestionValue={[Function (anonymous)]} renderSuggestion={[Function (anonymous)]} inputProps={{...}} renderSuggestionsContainer={[Function: defaultRenderSuggestionsContainer]} shouldRenderSuggestions={[Function: defaultShouldRenderSuggestions]} alwaysRenderSuggestions={false} multiSection={false} focusInputOnSuggestionClick={true} highlightFirstSuggestion={false} theme={{...}} id=\\"1\\">
                        <Autowhatever multiSection={false} items={{...}} renderInputComponent={[Function: defaultRenderInputComponent]} renderItemsContainer={[Function (anonymous)]} renderItem={[Function (anonymous)]} renderItemData={{...}} renderSectionTitle={[Function: renderSectionTitle]} getSectionItems={[Function: getSectionItems]} highlightedSectionIndex={{...}} highlightedItemIndex={{...}} inputProps={{...}} itemProps={[Function (anonymous)]} theme={{...}} id=\\"1\\" containerProps={{...}}>
                          <div role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"react-autowhatever-1\\" aria-expanded={false} className=\\"react-autosuggest__container\\">
                            <input type=\\"text\\" value=\\"\\" autoComplete=\\"off\\" aria-autocomplete=\\"list\\" aria-controls=\\"react-autowhatever-1\\" aria-activedescendant={{...}} className=\\"react-autosuggest__input\\" onChange={[Function: onChange]} onClick={[Function (anonymous)]} onFocus={[Function (anonymous)]} onBlur={[Function (anonymous)]} maxLength={22} onKeyDown={[Function (anonymous)]} />
                            <div id=\\"react-autowhatever-1\\" role=\\"listbox\\" className=\\"react-autosuggest__suggestions-container\\" />
                          </div>
                        </Autowhatever>
                      </Autosuggest>
                    </div>
                    <button id=\\"command-enter\\" aria-label=\\"Enter\\" className=\\"jsx-2334938183 style-scope ytd-searchbox\\">
                      <i className=\\"jsx-2334938183 fas fa-arrow-right\\" />
                    </button>
                    <JSXStyle id=\\"1731884614\\" dynamic={{...}} />
                  </form>
                </CommandBarInput>
                <JSXStyle id=\\"2638007119\\" />
              </div>
            </NoSSRCommandBar>
            <NavMenu>
              <NoSSR>
                <NoSSRNavMenu router={{...}}>
                  <nav className=\\"jsx-3914632746 container\\">
                    <div onClick={[Function (anonymous)]} className=\\"jsx-3914632746 return-btn\\">
                      <i className=\\"jsx-3914632746 fas fa-chevron-left\\" />
                    </div>
                    <a href=\\"/\\" className=\\"navmenu-link\\">
                      /
                    </a>
                    <a href=\\"/test\\" className=\\"navmenu-link\\">
                      <span>
                         
                      </span>
                      test
                      /
                    </a>
                    <JSXStyle id=\\"1346168239\\" dynamic={{...}} />
                  </nav>
                </NoSSRNavMenu>
              </NoSSR>
            </NavMenu>
          </div>
          <div className=\\"jsx-3778838050 mobile\\">
            <NoSSRMobileMenu router={{...}}>
              <ButtonsBar menuTexts={{...}} activeIndex={0}>
                <ul className=\\"jsx-4214066626 btnsbar with-indicator\\">
                  <li className=\\"btnsbar-item is-active\\">
                    <Link href=\\"/\\">
                      <a onClick={[Function: onClick]} onMouseEnter={[Function (anonymous)]} href=\\"/\\">
                        Home
                      </a>
                    </Link>
                  </li>
                  <li className=\\"btnsbar-item\\">
                    <Link href=\\"/about\\">
                      <a onClick={[Function: onClick]} onMouseEnter={[Function (anonymous)]} href=\\"/about\\">
                        About
                      </a>
                    </Link>
                  </li>
                  <li className=\\"btnsbar-item\\">
                    <Link href=\\"/creation\\">
                      <a onClick={[Function: onClick]} onMouseEnter={[Function (anonymous)]} href=\\"/creation\\">
                        Showcase
                      </a>
                    </Link>
                  </li>
                </ul>
                <JSXStyle id=\\"3650206365\\" dynamic={{...}} />
              </ButtonsBar>
            </NoSSRMobileMenu>
          </div>
          <JSXStyle id=\\"3778838050\\" />
        </div>
      </CommandBarNoSSR>
    </Suspense>
  </NoSSR>
</Memo(CommandBar)>"
`;
